[Proposal] Add the 'Super call enforcement' proposal (#1167)
apple/swift-evolution: proposals/xxxx-super-call-enforcement.md

# Super call enforcement

* Proposal: [SE-XXXX](xxxx-super-call-enforcement.md)
* Author: [Suyash Srijan](https://github.com/theblixguy)
* Review Manager: **TBD**
* Status: **Pending Review**
* Bug: [SR-6417](https://bugs.swift.org/browse/SR-6417)
* Implementation: [apple/swift#32712](https://github.com/apple/swift/pull/32712)

> * Toolchain: [macOS](https://ci.swift.org/job/swift-PR-toolchain-osx/604/artifact/branch-master/swift-PR-32712-604-osx.tar.gz) and [Linux](https://ci.swift.org/job/swift-PR-toolchain-Linux/426/artifact/branch-master/swift-PR-32712-426-ubuntu16.04.tar.gz)

Reviewer: These toolchains have expired (404 Not Found).

```suggestion
```

# Super call enforcement

* Proposal: [SE-XXXX](xxxx-super-call-enforcement.md)
* Author: [Suyash Srijan](https://github.com/theblixguy)
* Review Manager: **TBD**
* Status: **Pending Review**
* Bug: [SR-6417](https://bugs.swift.org/browse/SR-6417)
* Implementation: [apple/swift#32712](https://github.com/apple/swift/pull/32712)
* Toolchain: [macOS](https://ci.swift.org/job/swift-PR-toolchain-osx/604/artifact/branch-master/swift-PR-32712-604-osx.tar.gz) and [Linux](https://ci.swift.org/job/swift-PR-toolchain-Linux/426/artifact/branch-master/swift-PR-32712-426-ubuntu16.04.tar.gz)

## Introduction

Introduce a new attribute to enforce that an overridden function should call the `super` method in its body.

Swift-evolution thread: [Super call enforcement](https://forums.swift.org/t/super-call-enforcement/38177)

## Motivation

It is quite common to override a method in order to add additional functionality to it, rather than to completely replace it. For example - one might override `UIViewController.awakeFromNib()` to add additional setup code to it:

```swift
class MyViewController: UIViewController {

>   override func awakeFromNib()() {

Reviewer: ```suggestion
  override func awakeFromNib() {
```

```swift
class C1 {
  @requiresSuper 
  func foo() {}
}

class C2: C1 {
  override func foo() {} // error: method override is missing 'super.foo()' call
}

class C3: C1 {
  override func foo() { // Okay
    super.foo()
  }
}
```

## Detailed Design

The `super` call must be made to the base method that is annotated with `@requiresSuper` and not to some other base method. For example:

```swift
class C1 {
  @requiresSuper
  func foo() {}
  func foo(arg: Bool) {}
  func foo(arg: Int = 3) {}
}

class C2: C1 {
  override func foo() { super.foo(arg: true) } // error: method override is missing 'super.foo()' call
}

class C3: C1 {
  override func foo() { super.foo(arg: 0) } // error: method override is missing 'super.foo()' call
}

class C4: C1 {
  override func foo() { super.foo() } // Okay (since this is calling 'foo()' and not 'foo(arg: Int = 0)')
}
```

A few more examples:

```swift
class C1 {
  @requiresSuper
  func foo(_ arg: Int) {}
  func foo(_ arg: Bool) {}
}

class C2: C1 {
  override func foo(_ arg: Int) { super.foo(true) } // error: method override is missing 'super.foo(_:)' call
}

class C3: C1 {
  override func foo(_ arg: Int) { super.foo(0) } // Okay
}
```

> An message can be (optionally) specified on the `@requiresSuper` attribute to provide any additional information, which will be shown with the error message:

Reviewer 2: ```suggestion
A message can be (optionally) specified on the `@requiresSuper` attribute to provide any additional information, which will be shown with the error message:
```

```swift
class C1 {
  @requiresSuper("Call super as the final step in your implementation")
  func foo() {}
}

class C2: C1 {
  override func foo() {} // error: method override is missing 'super.foo()' call: Call super as the final step in your implementation
}
```

Any overrides of a method annotated with `@requiresSuper` will also implicitly inherit the attribute, to make sure that all the overrides in a subclass chain call back to the `super` method. This will be suppressed if the override is in a `final` class:

```swift
class C1 {
  @requiresSuper
  func foo() {}
}

class C2: C1 {
  // Implicitly inherits '@requiresSuper' from 'C1.foo'
  override func foo() {
    super.foo()
    // Do some other stuff
  }
}

final class C3: C1 {

>   // Does not implicitly inherits '@requiresSuper' from 'C1.foo', since 'C3.foo' cannot be overridden

Reviewer 2: ```suggestion
  // Does not implicitly inherit '@requiresSuper' from 'C1.foo', since 'C3.foo' cannot be overridden
```

  override func foo {
    super.foo()
    // Do some other stuff
  }
}

class C4: C2 {
  override func foo() {} // error: method override is missing 'super.foo()' call
}
```

> Finally, a `@requiresSuper` attribute will also be implicitly added for imported ObjC methods which have been annotated with `objc_requires_super`. This enables warnings (instead of errors) for any method with such an attribute, such as `UIViewController.viewDidLoad()`:

Reviewer: `UIViewController.viewDidLoad()` doesn't appear to have the `NS_REQUIRES_SUPER` attribute.

Examples from UIKit:

* [`UIView.updateConstraints()`](https://developer.apple.com/documentation/uikit/uiview/1622512-updateconstraints)
  > Call [super updateConstraints] as the final step in your implementation.

* [`NSObject.awakeFromNib()`](https://developer.apple.com/documentation/objectivec/nsobject/1402907-awakefromnib)
  > You must call the super implementation of awakeFromNib to give parent classes the opportunity to perform any additional initialization they require. Although the default implementation of this method does nothing, many UIKit classes provide non-empty implementations. You may call the super implementation at any point during your own awakeFromNib method.

```swift
import UIKit

class MyViewController1: UIViewController {
  override func viewDidLoad() { } // warning: method override is missing 'super.viewDidLoad()' call
}

class MyViewController2: UIViewController {
  override func viewDidLoad() { super.viewDidLoad() } // Okay
}
```

## Source compatibility

This is an additive change, so it does not break any existing Swift code. An imported ObjC declaration annotated with `objc_requires_super` will trigger a warning (instead of an error) if the overridden method skips the `super` call, which helps preserve source compatibility while giving users time to update their code.

## Effect on ABI stability

This change has no effect on the ABI.

## Effect on API resilience

Adding `@requiresSuper` to any existing API method can break code if an override of that method does not already call `super` in its body. Removing the attribute does not have any effect.

## Alternatives considered

- Do not diagnose missing `super` calls.

> - Diagnose missing `super` calls but also enforce their placement: The ordering of a 'super' call (first, middle, last, etc) is not a syntactic requirement. Adding such a requirement will complicate API design and usage and will also not be possible to prove statically in some scenarios. For simplicitly, it would be better to only dictate that a superclass method is called at some point, rather than also dictating its order and how many times its called.

Reviewer: ```suggestion
- Diagnose missing `super` calls but also enforce their placement: The ordering of a `super` call (first, middle, last, etc) is not a syntactic requirement. Adding such a requirement will complicate API design and usage and will also not be possible to prove statically in some scenarios. For simplicity, it would be better to only dictate that a superclass method is called at some point, rather than also dictating its order and how many times it's called.
```

> - Diagnose missing 'super' calls but also enforce its placement: The ordering of a 'super' call (first, middle, last, etc) is not syntactical requirement. Adding such a requirement will complicate API design and usage and will also not be possible to prove statically in some scenarios. For simplicitly, it would be better to only dictate that a superclass method is called at some point, rather than also dictating its order and how many times its called.

Reviewer 2: ```suggestion
- Diagnose missing 'super' calls but also enforce their placement: The ordering of a 'super' call (first, middle, last, etc) is not a syntactic requirement. Adding such a requirement will complicate API design and usage and will also not be possible to prove statically in some scenarios. For simplicitly, it would be better to only dictate that a superclass method is called at some point, rather than also dictating its order and how many times its called.
```

> - Diagnose missing 'super' calls but also allow a way to opt-out using a `@ignoresSuper` attribute. It would be strange allow users to bypass something that has deemed as a "requirement" by an API author.

Reviewer 2: ```suggestion
- Diagnose missing 'super' calls but also allow a way to opt-out using a `@ignoresSuper` attribute. It would be strange to allow users to bypass something that has deemed as a "requirement" by an API author.
```

> - Diagnose missing `super` calls but also allow a way to opt-out using a `@ignoresSuper` attribute. It would be strange to allow users to bypass something that has deemed as a "requirement" by an API author.

Reviewer: ```suggestion
- Diagnose missing `super` calls but also allow a way to opt-out using an `@ignoresSuper` attribute. It would be strange to allow users to bypass something that was deemed as a "requirement" by an API author.
```

## Future Directions

> We could make the `super` call check the default and get rid of the `@requiresSuper` attribute. This would be source-breaking in practice, but can be mitigated by downgrading the error to a warning in a compatibility mode. We could offer a `@ignoresSuper` attribute instead to silence the diagnostic.

Reviewer: ```suggestion
We could make the `super` call check the default and get rid of the `@requiresSuper` attribute. This would be source-breaking in practice, but can be mitigated by downgrading the error to a warning in a compatibility mode. We could offer an `@ignoresSuper` attribute instead to silence the diagnostic.
```
